### PR TITLE
Enable comparing fat pointers

### PR DESCRIPTION
**Pull request by RalfJung**
Also refactor our binops a bit to make that happen more easily.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/operator.rs [hunk 1]
```diff
@@ -272,83 +272,70 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
...
-                    right_layout.ty.is_fn());
+                assert!(left.layout.ty.is_integral() || left.layout.ty.is_unsafe_ptr() ||
+                    left.layout.ty.is_fn());
+                assert!(right.layout.ty.is_integral() || right.layout.ty.is_unsafe_ptr() ||
+                    right.layout.ty.is_fn());
 
                 // Handle operations that support pointer values
-                if left.is_ptr() || right.is_ptr() || bin_op == mir::BinOp::Offset {
-                    return M::ptr_op(self, bin_op, left, left_layout, right, right_layout);
+                if left.to_scalar_ptr()?.is_ptr() ||
```
Comment (oli-obk) [inline src/librustc_mir/interpret/operator.rs]: isn't this problematic for miri? a fat pointer will cause an error here, will it not?

File: src/librustc_mir/interpret/place.rs [hunk 2]
```diff
@@ -487,9 +487,9 @@ where
             Deref => self.deref_operand(base.into())?,
 
             Index(local) => {
-                let n = *self.frame().locals[local].access()?;
-                let n_layout = self.layout_of(self.tcx.types.usize)?;
-                let n = self.read_scalar(OpTy { op: n, layout: n_layout })?;
+                let layout =  self.layout_of(self.tcx.types.usize)?;
```
Comment (oli-obk) [inline src/librustc_mir/interpret/place.rs]:
```suggestion
                let layout = self.layout_of(self.tcx.types.usize)?;
```

File: src/librustc_mir/interpret/operator.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/operator.rs]: `to_scalar_ptr` returns the first component of a scalar pair.

**Comment by RalfJung** [comment]
Rebased.

@bors r=oli-obk

**Comment by RalfJung** [comment]
Rebased, to integrate with other Miri-related changes (for easier testing).

@bors r=oli-obk

**Comment by RalfJung** [comment]
Fixed rebase fallout.

@bors r=oli-obk

**Pull request merged**